scipy/stats/_qmc.py

```diff
@@ -883,6 +977,177 @@ def reset(self):
+       optimization of Latin Hypercube Samples and subprojection properties",
+       Journal of Simulation, 2013.
+
+    Examples
+    --------
+    Generate samples from an optimal design.
+
+    >>> from scipy.stats import qmc
+    >>> sampler = qmc.OptimalDesign(d=2, seed=12345)
+    >>> sample = sampler.random(n=5)
+    >>> sample
+    array([[0.84052691, 0.53664833],
+           [0.1545328 , 0.36265316],
+           [0.52177809, 0.93343721],
+           [0.68033825, 0.06474907],
+           [0.26544879, 0.61163943]])
+
+    Compute the quality of the sample using the discrepancy criterion.
+
+    >>> qmc.discrepancy(sample)
+    0.01602810655510689
+
+    You can possibly improve the quality of the sample by performing more
+    optimization iterations by using `niter`:
+
+    >>> sampler_2 = qmc.OptimalDesign(d=2, niter=5, seed=12345)
+    >>> sample_2 = sampler_2.random(n=5)
+    >>> qmc.discrepancy(sample_2)
+    0.01600398742881648
+
+    """
+
+    def __init__(self, d, start_design=None, niter=1, method=None, seed=None):
+        super().__init__(d=d, seed=seed)
+        self.start_design = start_design
+        self.niter = niter
+
+        if method is None:
+            def method(func, x0, bounds):
+                """Basinhopping optimization."""
+                minimizer_kwargs = {"method": "L-BFGS-B", "bounds": bounds}
+                basinhopping(func, x0, niter=100,
+                             minimizer_kwargs=minimizer_kwargs, seed=self.rng)
+
+        self.method = method
+
+        self.best_doe = self.start_design
+        if self.start_design is not None:
+            self.best_disc = discrepancy(self.start_design)
+        else:
+            self.best_disc = np.inf
+
+        self.lhs = LatinHypercube(self.d, seed=self.rng)
+
+    def random(self, n=1):
+        """Draw `n` in the half-open interval ``[0, 1)``.
+
+        Parameters
+        ----------
+        n : int, optional
+            Number of samples to generate in the parameter space. Default is 1.
+
+        Returns
+        -------
+        sample : array_like (n, d)
+            Optimal sample.
+
+        """
+        if self.d == 0 or n == 0:
+            return np.empty((n, self.d))
+
+        if self.best_doe is None:
+            self.best_doe = self.lhs.random(n)
+            self.best_disc = discrepancy(self.best_doe)
+
+        if n == 1:
+            return self.best_doe
+
+        def _perturb_best_doe(x: np.ndarray) -> float:
+            """Perturb the DoE and keep track of the best DoE.
+
+            Parameters
+            ----------
+            x : list of int
+                It is a list of:
+                    idx : int
+                        Index value of the components to compute
+
+            Returns
+            -------
+            discrepancy : float
+                Centered discrepancy.
+
+            """
+            # Perturb the DoE
+            doe = self.best_doe.copy()
```

Comment:
This is going to be really expensive and it shouldn't be needed because `_perturb_discrepancy` doesn't actually need to do the permutation to calculate the new discrepancy, right? Even if it did, it would be faster to un-permute (if thepermutation is not an improvement) than to copy everything every iteration.